### PR TITLE
feat(core): add indexed color (0-255) and OSC palette control

### PR DESCRIPTION
**Pull request by bresilla**
## Summary

Adds full support for terminal indexed colors (0-255) and OSC-based terminal palette control, enabling the use of the standard 256-color palette and dynamic color modification via ANSI/OSC escape sequences.

Makes this possible (with tools like pywal et al):
ezgif-54dccc5a972ae0ee

### TypeScript (RGBA.ts)
- Added `COLOR_TYPE_RGB`, `COLOR_TYPE_INDEXED`, `COLOR_TYPE_DEFAULT` constants
- Extended `RGBA` class with `colorType`, `index` properties
- Added `RGBA.fromIndex(index)`, `RGBA.defaultForeground()`, `RGBA.defaultBackground()` static methods
- Added `isIndexed()`, `isDefault()`, `isRgb()` instance methods
- Added `indexToApproximateRgb()` function with standard 256-color palette mapping
- Extended `ColorInput` type to accept `string | RGBA | IndexedColor | number`
- Updated `parseColor()` to handle `"ansi:N"` format, numeric indices, and `{index: N}` objects

### TypeScript (terminal-palette.ts) - OSC Palette Control
- Added `setPaletteColor(index, hex)` - dynamically set any of the 256 palette colors (OSC 4)
- Added `setForeground(hex)` - set terminal default foreground color (OSC 10)
- Added `setBackground(hex)` - set terminal default background color (OSC 11)
- Added `setCursorColor(hex)` - set cursor color (OSC 12)
- Added `resetPaletteColor(index)` - reset palette color to terminal default (OSC 104)
- Added `resetForeground()` - reset foreground to default (OSC 110)
- Added `resetBackground()` - reset background to default (OSC 111)
- Added `resetCursorColor()` - reset cursor color to default (OSC 112)

### Zig (ansi.zig)
- Added `ColorType` enum: `rgb`, `indexed`, `default`
- Added `Color` struct combining RGBA with type and index
- Added `indexToApproximateRgba()`, `fgIndexedColorOutput()`, `bgIndexedColorOutput()`
- Added `fgDefaultOutput()`, `bgDefaultOutput()` for default color reset

### Zig (buffer.zig)
- Extended `Cell` struct with `fg_color_type`, `bg_color_type`, `fg_index`, `bg_index`
- Added corresponding arrays to `OptimizedBuffer`
- Updated all buffer operations to handle color types

### Zig (renderer.zig)
- Updated color output to check cell color type
- Outputs `\x1b[38;5;Nm` / `\x1b[48;5;Nm` for indexed colors
- Outputs `\x1b[39m` / `\x1b[49m` for default colors
- Falls back to RGB for rgb color type

### Indexed Colors
```typescript
import { parseColor, RGBA } from "@opentui/core"

// Various ways to specify indexed colors
parseColor("ansi:196")      // String format (bright red)
parseColor(46)              // Numeric (green)
parseColor({ index: 226 })  // Object (yellow)
RGBA.fromIndex(21)          // Direct method (blue)

// Check color type
const color = parseColor("ansi:196")
color.isIndexed()  // true
color.index        // 196
```

### OSC Palette Control
```typescript
// Dynamically modify terminal colors
paletteDetector.setPaletteColor(1, "#ff5555")  // Change color 1 to custom red
paletteDetector.setBackground("#1a1a2e")       // Set terminal background
paletteDetector.setForeground("#eaeaea")       // Set terminal foreground
paletteDetector.setCursorColor("#ff79c6")      // Set cursor color

// Reset to terminal defaults
paletteDetector.resetPaletteColor(1)           // Reset color 1
paletteDetector.resetBackground()              // Reset background
paletteDetector.resetForeground()              // Reset foreground
```

## Testing

- All 3057 existing tests pass
- Build completes successfully

### CODE REVIEW

File: packages/core/src/lib/terminal-palette.ts [hunk 1]
```diff
@@ -334,6 +342,52 @@ export class TerminalPalette implements TerminalPaletteDetector {
       highlightForeground: specialColors[19],
     }
   }
+
+  private normalizeHex(hex: string): string {
+    let h = hex.startsWith("#") ? hex.slice(1) : hex
+    if (h.length === 3) {
+      h = h[0] + h[0] + h[1] + h[1] + h[2] + h[2]
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/terminal-palette.ts]: Fixed  d907260 normalizeHex now validates hex input to avoid emitting malformed OSC sequences.

File: packages/core/src/lib/RGBA.ts [hunk 2]
```diff
@@ -62,11 +99,63 @@ export class RGBA {
...
+    if (this.colorType === COLOR_TYPE_DEFAULT) {
+      return "default"
+    }
     return `rgba(${this.r.toFixed(2)}, ${this.g.toFixed(2)}, ${this.b.toFixed(2)}, ${this.a.toFixed(2)})`
   }
 }
 
-export type ColorInput = string | RGBA
+export type IndexedColor = { index: number }
+export type ColorInput = string | RGBA | IndexedColor | number
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Intentional API choice: numeric input maps to indexed color (0–255) for ergonomics; explicit formats (ansi:N, { index }) remain supported.

File: packages/core/src/lib/RGBA.ts [hunk 3]
```diff
@@ -182,18 +271,51 @@ const CSS_COLOR_NAMES: Record<string, string> = {
...
   if (typeof color === "string") {
     const lowerColor = color.toLowerCase()
 
     if (lowerColor === "transparent") {
       return RGBA.fromValues(0, 0, 0, 0)
     }
 
+    if (lowerColor === "default") {
+      return RGBA.defaultForeground()
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Fixed: parseColor now accepts fg/bg context, and call sites use parseColor(..., \"bg\") where appropriate.

File: packages/core/src/lib/RGBA.ts [hunk 4]
```diff
@@ -1,8 +1,16 @@
...
 export class RGBA {
   buffer: Float32Array
+  colorType: number
+  index: number
 
-  constructor(buffer: Float32Array) {
+  constructor(buffer: Float32Array, colorType: number = COLOR_TYPE_RGB, index: number = 0) {
     this.buffer = buffer
+    this.colorType = colorType
+    this.index = index
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Fixed across commits d907260, 222ab4b: FFI now passes color type/index; styled text, boxes, and renderer background preserve indexed/default semantics.

File: packages/core/src/zig/buffer.zig [hunk 5]
```diff
@@ -769,6 +847,60 @@ pub const OptimizedBuffer = struct {
...
+            .bg_index = bg_index,
+        };
+
+        if (self.get(x, y)) |destCell| {
+            const blendedCell = blendCells(overlayCell, destCell);
+            self.set(x, y, blendedCell);
+        } else {
+            self.set(x, y, overlayCell);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/zig/buffer.zig]: Fixed: colorType-aware setters are now used via FFI and box/textbuffer rendering paths (d907260, 222ab4b).

File: packages/core/src/lib/RGBA.ts [hunk 6]
```diff
@@ -21,6 +29,35 @@ export class RGBA {
...
+    return this.colorType === COLOR_TYPE_INDEXED
+  }
+
+  isDefault(): boolean {
+    return this.colorType === COLOR_TYPE_DEFAULT
+  }
+
+  isRgb(): boolean {
+    return this.colorType === COLOR_TYPE_RGB
+  }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts:64]: Fixed in 0eeea68: added RGBA/fromIndex + parseColor numeric/object + default fg/bg tests.

File: packages/core/src/lib/terminal-palette.ts [hunk 7]
```diff
@@ -334,6 +342,52 @@ export class TerminalPalette implements TerminalPaletteDetector {
...
+    this.writeOsc(`\x1b]110\x07`)
+  }
+
+  resetBackground(): void {
+    this.writeOsc(`\x1b]111\x07`)
+  }
+
+  resetCursorColor(): void {
+    this.writeOsc(`\x1b]112\x07`)
+  }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/terminal-palette.ts]: Fixed in d907260: added tests verifying emitted OSC sequences

File: packages/core/src/zig/ansi.zig [hunk 8]
```diff
@@ -3,6 +3,88 @@ const Allocator = std.mem.Allocator;
...
+
+    pub fn isDefault(self: Color) bool {
+        return self.color_type == .default;
+    }
+
+    pub fn isRgb(self: Color) bool {
+        return self.color_type == .rgb;
+    }
+};
+
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/zig/ansi.zig]: Fixed (and this comment is outdated): removed unused Color struct in d907260.

File: packages/core/src/lib/RGBA.test.ts [hunk 9]
```diff
@@ -726,6 +726,23 @@ describe("parseColor", () => {
...
+  test("parses default keyword as fg by default and bg when context is bg", () => {
+    const fg = parseColor("default")
+    const bg = parseColor("default", "bg")
+
+    expect(fg.isDefault()).toBe(true)
+    expect(bg.isDefault()).toBe(true)
+
+    // In bg context, the default background constructor uses alpha 0 to avoid forcing a color
+    expect(bg.a).toBe(0)
+  })
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.test.ts]: Fixed in 0eeea68: added explicit tests for RGBA.defaultForeground() and RGBA.defaultBackground().

File: packages/core/src/lib/RGBA.test.ts [hunk 10]
```diff
@@ -726,6 +726,23 @@ describe("parseColor", () => {
...
+
+  test("parses ansi:N format as indexed color", () => {
+    const rgba = parseColor("ansi:196")
+    expect(rgba.isIndexed()).toBe(true)
+    expect(rgba.index).toBe(196)
+  })
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.test.ts]: Fixed in 0eeea68: added coverage for { index } input.

File: packages/core/src/lib/terminal-palette.ts [hunk 11]
```diff
@@ -334,6 +342,66 @@ export class TerminalPalette implements TerminalPaletteDetector {
...
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/terminal-palette.ts]: Not adding JSDoc because project TS style guidelines say no JSDoc. If maintainers want docs, happy to add a short section to docs/README instead.

File: packages/core/src/lib/RGBA.ts [hunk 12]
```diff
@@ -21,6 +29,35 @@ export class RGBA {
...
   }
 
+  static fromIndex(index: number): RGBA {
+    if (index < 0 || index > 255) {
+      console.warn(`Invalid indexed color: ${index}, must be 0-255, defaulting to 0`)
+      index = 0
+    }
+    const rgb = indexToApproximateRgb(index)
+    return new RGBA(new Float32Array([rgb.r, rgb.g, rgb.b, 1.0]), COLOR_TYPE_INDEXED, index)
+  }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: This is already validated before approximation: index is clamped to 0..255 before calling indexToApproximateRgb(index).

File: packages/core/src/lib/RGBA.ts [hunk 13]
```diff
@@ -181,19 +270,56 @@ const CSS_COLOR_NAMES: Record<string, string> = {
...
+    // Handle "ansi:N" format for indexed colors
+    if (lowerColor.startsWith("ansi:")) {
+      const indexStr = lowerColor.slice(5)
+      const index = parseInt(indexStr, 10)
+      if (!isNaN(index)) {
+        return RGBA.fromIndex(index)
+      }
+      console.warn(`Invalid ansi color format: ${color}, defaulting to magenta`)
+      return RGBA.fromValues(1, 0, 1, 1)
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Out-of-range indices are handled safely by RGBA.fromIndex (warn + clamp to 0). We kept that behavior instead of returning magenta.

File: packages/core/src/lib/RGBA.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.test.ts]: Fixed in 0eeea68: added parseColor(46) test

[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/RGBA.test.ts]: Fixed in 0eeea68: added fromIndex tests for valid + invalid indices.

File: packages/core/src/buffer.ts [hunk 14]
```diff
@@ -155,12 +155,23 @@ export class OptimizedBuffer {
...
   public setCell(x: number, y: number, char: string, fg: RGBA, bg: RGBA, attributes: number = 0): void {
     this.guard()
-    this.lib.bufferSetCell(this.bufferPtr, x, y, char, fg, bg, attributes)
+    // Check if either color has color type info (indexed or default)
+    const hasColorTypeInfo = fg.colorType !== undefined || bg.colorType !== undefined
+    if (hasColorTypeInfo) {
+      this.lib.bufferSetCellWithColorType(this.bufferPtr, x, y, char, fg, bg, attributes)
+    } else {
+      this.lib.bufferSetCell(this.bufferPtr, x, y, char, fg, bg, attributes)
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/buffer.ts]: Fixed in 0eeea68

File: packages/core/src/buffer.ts [hunk 15]
```diff
@@ -172,7 +183,13 @@ export class OptimizedBuffer {
...
   ): void {
     this.guard()
-    this.lib.bufferSetCellWithAlphaBlending(this.bufferPtr, x, y, char, fg, bg, attributes)
+    // Check if either color has color type info (indexed or default)
+    const hasColorTypeInfo = fg.colorType !== undefined || bg.colorType !== undefined
+    if (hasColorTypeInfo) {
+      this.lib.bufferSetCellWithAlphaBlendingAndColorType(this.bufferPtr, x, y, char, fg, bg, attributes)
+    } else {
+      this.lib.bufferSetCellWithAlphaBlending(this.bufferPtr, x, y, char, fg, bg, attributes)
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/buffer.ts]: Fixed in 0eeea68

File: packages/core/src/lib/terminal-palette.ts [hunk 16]
```diff
@@ -334,6 +342,66 @@ export class TerminalPalette implements TerminalPaletteDetector {
...
+
+  setCursorColor(hex: string): void {
+    const h = this.normalizeHex(hex)
+    this.writeOsc(`\x1b]12;#${h}\x07`)
+  }
+
+  resetPaletteColor(index: number): void {
+    if (index < 0 || index > 255) return
+    this.writeOsc(`\x1b]104;${index}\x07`)
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/lib/terminal-palette.ts]: Good catch. I changed setPaletteColor/resetPaletteColor to throw on out-of-range indices for consistency with validation behavior.

File: packages/core/src/buffer.ts [hunk 17]
```diff
@@ -155,12 +155,23 @@ export class OptimizedBuffer {
 
   public clear(bg: RGBA = RGBA.fromValues(0, 0, 0, 1)): void {
     this.guard()
-    this.lib.bufferClear(this.bufferPtr, bg)
+    const hasColorTypeInfo = bg.colorType !== undefined
+    if (hasColorTypeInfo) {
+      this.lib.bufferClearWithColorType(this.bufferPtr, bg)
+    } else {
+      this.lib.bufferClear(this.bufferPtr, bg)
+    }
```
[the remark this answers is not in the document]
Comment (bresilla) [reply packages/core/src/buffer.ts]: Fixed in 0eeea68 (fix(core): sanitize color metadata for FFI)

File: packages/core/src/lib/RGBA.ts [hunk 18]
```diff
@@ -1,8 +1,16 @@
+export const COLOR_TYPE_RGB = 0
```
Comment (kommander) [inline packages/core/src/lib/RGBA.ts]: I would make this an enum.

File: packages/core/src/lib/RGBA.ts [hunk 19]
```diff
@@ -181,19 +270,56 @@ const CSS_COLOR_NAMES: Record<string, string> = {
   brightwhite: "#FFFFFF",
 }
 
-export function parseColor(color: ColorInput): RGBA {
+export function parseColor(color: ColorInput, context: "fg" | "bg" = "fg"): RGBA {
```
Comment (kommander) [inline packages/core/src/lib/RGBA.ts]: `parseColor` is used in so many places... I currently don't think it should be context aware.

File: packages/core/src/lib/terminal-palette.ts [hunk 20]
```diff
@@ -30,6 +30,14 @@ export interface TerminalPaletteDetector {
   detect(options?: GetPaletteOptions): Promise<TerminalColors>
   detectOSCSupport(timeoutMs?: number): Promise<boolean>
   cleanup(): void
+  setPaletteColor(index: number, hex: string): void
+  setForeground(hex: string): void
+  setBackground(hex: string): void
+  setCursorColor(hex: string): void
```
Comment (kommander) [inline packages/core/src/lib/terminal-palette.ts]: There is a method on `renderer` already to set the cursor color I think. That would be duplicated.

**Comment by kommander** [comment]
Thanks for taking a stab at this. Having `RGBA` with additional attributes for `type` and `index` it adds some overhead in passing it around. When I was thinking about how to do this I thought it could maybe be added as additional `f32`, making it `RGBA+Attributes`, where the last `f32` can be used to encode type + index and use bitmasks to encode/decode. I would make framebuffers use slightly more memory, but most apps don't use dozens of large framebuffers, so mostly negligible. 

I'd like to try and keep the API surface impact as small as possible, like keeping `parseColor` stateless and not context aware.

File: packages/core/src/lib/RGBA.ts (hunk 19 above)
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Fixed: parseColor is stateless again. Added parseFgColor() / parseBgColor() helpers so only callers that need bg semantics opt in, while the default API remains unchanged.

File: packages/core/src/lib/terminal-palette.ts (hunk 20 above)
Comment (bresilla) [reply packages/core/src/lib/terminal-palette.ts]: Good call — agreed. The renderer already exposes cursor color control (CliRenderer.setCursorColor(...)), so I removed setCursorColor/resetCursorColor from TerminalPalette (and updated the related tests) to avoid duplicating the API surface.

File: packages/core/src/lib/RGBA.ts (hunk 18 above)
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: Agreed an enum would be cleaner. I didn’t change it in this PR because those values cross the TS↔Zig FFI boundary (0/1/2) and we’re already touching a lot of call sites/ABI surfaces here; keeping plain numeric constants reduced churn and made it easier to verify we stayed aligned with Zig. If you’d like, I can do a small follow-up commit that introduces a ColorType enum (backed by the same 0/1/2 values) and updates the internal TS usages without changing the FFI encoding.

**Comment by bresilla** [comment]
> Thanks for taking a stab at this. Having `RGBA` with additional attributes for `type` and `index` it adds some overhead in passing it around. When I was thinking about how to do this I thought it could maybe be added as additional `f32`, making it `RGBA+Attributes`, where the last `f32` can be used to encode type + index and use bitmasks to encode/decode. I would make framebuffers use slightly more memory, but most apps don't use dozens of large framebuffers, so mostly negligible.
> 
> I'd like to try and keep the API surface impact as small as possible, like keeping `parseColor` stateless and not context aware.

I updated things so parseColor is stateless again. Callers that need “default background” semantics now opt into `parseBgColor()` (and `parseFgColor()` for symmetry), so we no longer have to thread a context parameter through the entire codebase.

On the representation: I agree that an RGBA + meta packed float (or similar) would be a nice improvement to reduce object metadata overhead and allow the color type/index to travel naturally with the buffer. That’s a larger refactor—it would touch FFI structs, buffer layout, and a fair bit of Zig/TS packing code—so I kept this PR focused on correctness and on wiring the metadata through all render paths.

I can follow up with a separate PR (maybe??) that explores the packed-f32 approach while keeping the external API stable.

**Comment by kommander** [comment]
It's still alpha, so I wouldn't mind breaking some APIs at this stage to get it "right". But also, what high level APIs would break? `RGBA` interfaces wouldn't have to change with defaults I think? And passing RGBA around should then basically be the same.

**Comment by bresilla** [comment]
> It's still alpha, so I wouldn't mind breaking some APIs at this stage to get it "right". But also, what high level APIs would break? `RGBA` interfaces wouldn't have to change with defaults I think? And passing `RGBA` around should then basically be the same.

Totally fair — if we’re willing to break some APIs now, doing the packed RGBA + meta is probably the cleanest long-term shape.
**But actually we don't have to have high-level API breakage - if we’re careful**:
- We can keep the public `RGBA` class surface the same (`fromValues`/`fromHex`/`fromIndex`/`defaultForeground`/`defaultBackground`, `isRgb`/`isIndexed`/`isDefault`, index, etc).
- Passing `RGBA` around in TS stays basically the same.

What would likely change is more “internal/FFI plumbing” than user-facing APIs:
- Any code that assumes RGBA.buffer is exactly 4 floats would need updating (it may become 5 floats or a different encoding).
- All Zig FFI entrypoints that take `[*]const f32` expecting `RGBA(4)` would need to accept the new layout (or we keep passing `RGBA(4)` plus separate `u8s`, as now).
- Any helper that directly manipulates framebuffer arrays (fg/bg buffers) might need to allocate/handle the extra metadata lane.

**Comment by bresilla** [comment]
I think this PR can merge as-is for correctness, and then if you’re happy with the direction, I can open a small follow-up PR that experiments with the packed RGBA+meta representation (bit-packing type/index into an extra float). The aim would be to keep the public `RGBA/parseColor` APIs the same and confine other bits to internal buffers + TS↔Zig plumbing.

**Comment by bresilla** [comment]
Looking at this deeper, "a safer phased plan” -so to say- is if we do it later as a follow-up PR with those stages (maybe!) - and is becoming a large PR as it is...:

Step 1 (no buffer layout change yet)
- Introduce a ColorType enum in TS + Zig helper functions to standardize meta.
- Keep .buffer at 4 floats, keep passing meta u8s separately (as we do now), just refactor for clarity.

Step2 (packed meta, but keep separate arrays)
- Make RGBA.buffer 5 floats and encode meta in float `#5`.
- In FFI, decode meta once at entrypoints and still fill fg_color_type/index arrays (so internal Zig logic stays mostly same).
- This isolates changes to FFI boundary + RGBA class, not the whole renderer.

Step 3 (maybe some optimization)
- Remove separate arrays and store packed meta directly in the buffer... (?)

**Comment by kommander** [comment]
That could make sense to do it step by step. @simonklee what do you think? 

@bresilla there is some segfault in the tests currently though.

**Comment by bresilla** [comment]
> That could make sense to do it step by step. @simonklee what do you think?
> 
> @bresilla there is some segfault in the tests currently though.

Sorry — I was initially only running core tests. Should be fixed now...

Anyway, Solid test use `SyntaxStyle.fromStyles` often and are called with string colors (e.g. fg: "#C792EA"). Normalizing `fg`/`bg` inside `registerStyle` keeps that API working while ensuring only `RGBA` reaches Zig, preventing runtime crashes and keeping internal types consistent.


Edit: still not passing, gonna check it later today.
Sorry for the mess 😅

Edit2: I think the error is not related to the code i added... I think. Sorry not very good at js/ts/bun things. But according to my understanding the error is about bun version bla bla bla... :P

**Review by simonklee — COMMENTED** [review]
In general its hard to know if all edge-cases have been covered because there are many of these draw paths. I walked through a couple and found a few that still bypass the new colorType/index
handling. I don't know that tests are good enough for us to catch all of these either.

File: packages/core/src/lib/RGBA.ts [hunk 21]
```diff
@@ -62,11 +99,63 @@ export class RGBA {
...
+    const b = cubeIndex % 6
+    const g = Math.floor(cubeIndex / 6) % 6
+    const r = Math.floor(cubeIndex / 36)
+    // 6x6x6 cube: values are 0, 95, 135, 175, 215, 255 -> normalized
+    const levels = [0, 0.373, 0.529, 0.686, 0.843, 1.0]
+    return { r: levels[r], g: levels[g], b: levels[b] }
+  }
+
+  // Grayscale (232-255)
+  const gray = (index - 232) * (1.0 / 23)
```
Comment (simonklee) [inline packages/core/src/lib/RGBA.ts]: xterm grayscale uses values 8, 18, ... 238 (steps of 10)
```
const gray = (8 + (index - 232) * 10) / 255
```

File: packages/core/src/zig/buffer.zig [hunk 22]
```diff
@@ -845,11 +1005,15 @@ pub const OptimizedBuffer = struct {
                 const rowSliceFg = self.buffer.fg[rowStartIndex .. rowStartIndex + rowWidth];
                 const rowSliceBg = self.buffer.bg[rowStartIndex .. rowStartIndex + rowWidth];
                 const rowSliceAttrs = self.buffer.attributes[rowStartIndex .. rowStartIndex + rowWidth];
+                const rowSliceBgType = self.buffer.bg_color_type[rowStartIndex .. rowStartIndex + rowWidth];
+                const rowSliceBgIndex = self.buffer.bg_index[rowStartIndex .. rowStartIndex + rowWidth];
```
Comment (simonklee) [inline packages/core/src/zig/buffer.zig:879]: Aren't we missing the fg meta in fast path here?

```
const rowSliceFgType = self.buffer.fg_color_type[rowStartIndex .. rowStartIndex + rowWidth];
const rowSliceFgIndex = self.buffer.fg_index[rowStartIndex .. rowStartIndex + rowWidth];
```

File: packages/core/src/lib/RGBA.ts (hunk 21 above)
Comment (bresilla) [reply packages/core/src/lib/RGBA.ts]: I was just following this: 

image


which is from [ANSI Escape Colors](https://en.wikipedia.org/wiki/ANSI_escape_code)

or a i misunderstanding something??

File: packages/core/src/zig/buffer.zig (hunk 22 above)
Comment (bresilla) [reply packages/core/src/zig/buffer.zig:879]: Yeah, sorry, mistake. Will add that :)

**Comment by bresilla** [comment]
> In general its hard to know if all edge-cases have been covered because there are many of these draw paths. I walked through a couple and found a few that still bypass the new colorType/index handling. I don't know that tests are good enough for us to catch all of these either.

I will try to `grep` and check as much as i can, TS is not my language though... soo it's a bit tricky to follow :P

File: packages/core/src/lib/RGBA.ts (hunk 21 above)
Comment (simonklee) [reply packages/core/src/lib/RGBA.ts]: 20260125_15h19m11s_grim
From the wiki article `level = gray * 10 + 8`

**Comment by kommander** [comment]
We should probably get some of the draw paths test covered more thoroughly and make sure those tests pass on main as well, so we can assure at least somewhat same behaviour.

**Comment by kommander** [comment]
@bresilla thank you for pushing this forward, I really want to have this. I think we should go through with integrating the color info into the RGBA+A bytes to pass around. I know it's a big undertaking, that's why it wasn't done yet. I get the idea of taking it step by step, but I think the way it's currently done it introduces complexity we have to clean up in the next step, while the first step can be clane and I don't like these spreading through the codebase when the RGBA+A type can encode and transport them as a single unit:

```diff
 pub const StyleDefinition = struct {
     fg: ?RGBA,
     bg: ?RGBA,
+    fg_color_type: ansi.ColorType = .rgb,
+    bg_color_type: ansi.ColorType = .rgb,
+    fg_index: u8 = 0,
+    bg_index: u8 = 0,
     attributes: u32,
 };
```

It should not be necessary to have additional methods like `bufferSetCellWithColorType` or `bufferSetCellWithAlphaBlendingAndColorType` to handle things, all existing methods should basically be enough and handle the new RGBA+A type appropriately.

**Comment by bresilla** [comment]
> @bresilla thank you for pushing this forward, I really want to have this. I think we should go through with integrating the color info into the RGBA+A bytes to pass around. I know it's a big undertaking, that's why it wasn't done yet. I get the idea of taking it step by step, but I think the way it's currently done it introduces complexity we have to clean up in the next step, while the first step can be clane and I don't like these spreading through the codebase when the RGBA+A type can encode and transport them as a single unit:
> 
> ```diff
>  pub const StyleDefinition = struct {
>      fg: ?RGBA,
>      bg: ?RGBA,
> +    fg_color_type: ansi.ColorType = .rgb,
> +    bg_color_type: ansi.ColorType = .rgb,
> +    fg_index: u8 = 0,
> +    bg_index: u8 = 0,
>      attributes: u32,
>  };
> ```
> 
> It should not be necessary to have additional methods like `bufferSetCellWithColorType` or `bufferSetCellWithAlphaBlendingAndColorType` to handle things, all existing methods should basically be enough and handle the new RGBA+A type appropriately.


Hey, sorry for late response.
No worries, I would like to give a try your way. I don't have time coming days, but I am hoping after that I can spend an evening or two on this.

How would you suggest me doing it? Close this and start from scratch? I think that would be the cleanest way...

**Comment by kommander** [comment]
I'll pick this up asap, if I get to it before you do I'll take your branch here as a base. I would just work from what we have here now.

**Comment by bresilla** [comment]
## Updated Implementation: Packed Metadata Approach

This PR has been reworked based on review feedback. Instead of adding separate `fg_color_type`, `bg_color_type`, `fg_index`, `bg_index` fields throughout the codebase, color metadata is now **packed into the RGBA type itself** as a 5th float.

### Encoding scheme

`meta = colorType * 256.0 + index` stored in `RGBA[4]`

| Color Type | meta | Example RGBA |
|---|---|---|
| RGB | `0.0` | `[0.8, 0.2, 0.3, 1.0, 0.0]` |
| Indexed (196) | `452.0` | `[0.8, 0.0, 0.0, 1.0, 452.0]` |
| Default | `512.0` | `[1.0, 1.0, 1.0, 1.0, 512.0]` |

Decode: `colorType = floor(meta / 256)`, `index = meta % 256`

### What changed

**Zig side:**
- `RGBA = [4]f32` → `RGBA = [5]f32`
- No new struct fields — meta travels with the color naturally
- Renderer dispatches on color type: indexed → `\x1b[38;5;Nm`, default → `\x1b[39m`, rgb → `\x1b[38;2;R;G;Bm`
- Blended colors reset meta to `0.0` (blending destroys indexed identity)
- All ~430 RGBA literals in tests updated

**TypeScript side:**
- `RGBA` class: `fromIndex()`, `defaultColor()`, `isIndexed()`, `isDefault()`, `isRgb()`, `meta`/`colorType`/`colorIndex` getters
- Buffer stride `* 4` → `* 5` across `buffer.ts`, `filters.ts`, and all test files
- FFI boundary updated (`zig-structs.ts`, `zig.ts`, `lib.zig`)

### Test results
- Zig: `zig build test` — all pass
- TypeScript: `bun test` — 3340 pass, 0 fail

**Comment by bresilla** [comment]
> I'll pick this up asap, if I get to it before you do I'll take your branch here as a base. I would just work from what we have here now.

@kommander i think i managed to do it (with the help of OC). Would you be able to review/look at it??

**Comment by bresilla** [comment]
> I'll pick this up asap, if I get to it before you do I'll take your branch here as a base. I would just work from what we have here now.

@kommander I know you must be super busy, but do you think you might have a chance in coming weeks to have a look at this?? :) 

Thank you in advance :)

**Comment by simonklee** [comment]
> > I'll pick this up asap, if I get to it before you do I'll take your branch here as a base. I would just work from what we have here now.
> 
> @kommander I know you must be super busy, but do you think you might have a chance in coming weeks to have a look at this?? :)
> 
> Thank you in advance :)

I'm having a look at it this at the moment.

**Comment by jahooma** [comment]
This is a really important fix to get in, as the colors look very broken in the Apple terminal, which is a large fraction of all Mac usage for us.

Questions:
- When can a fix be merged?
- Are there any workarounds for now?

**Comment by simonklee** [comment]
> This is a really important fix to get in, as the colors look very broken in the Apple terminal, which is a large fraction of all Mac usage for us.
> 
> Questions:
> 
> * When can a fix be merged?
> * Are there any workarounds for now?

I haven an open pr that adopts this and we intend to merge it once reviewed. https://github.com/anomalyco/opentui/pull/844

You can't really work around it.

**Comment by simonklee** [comment]
https://github.com/anomalyco/opentui/commit/95d36f3553ea9eb2b3a5f4d9f13ecd261e51602d
